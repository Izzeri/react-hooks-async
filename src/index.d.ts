export type DependencyList = ReadonlyArray<unknown>;

export type AsyncTask<Result> = {
  started: boolean;
  pending: boolean;
  error: Error | null;
  result: Result | null;
  start: () => void | null;
  abort: () => void | null;
};

export type UseAsyncTask = <Result>(
  func: (c: AbortController) => Promise<Result>,
  deps: DependencyList,
) => AsyncTask<Result>;

type Falsy = false | 0 | '' | null | undefined;
export type UseAsyncRun = (t: AsyncTask<unknown> | Falsy) => void;

export type UseAsyncCombine = (
  ...ts: AsyncTask<unknown>[],
) => AsyncTask<unknown> & { errorAll?: Error[] };

// core async hooks

export const useAsyncTask: UseAsyncTask;
export const useAsyncRun: UseAsyncRun;

// combining async hooks

export const useAsyncCombineAll: UseAsyncCombine;
export const useAsyncCombineSeq: UseAsyncCombine;
export const useAsyncCombineRace: UseAsyncCombine;

// timeout task

export type UseAsyncTaskTimeout = <Result>(
  func: () => Result,
  delay: number,
  deps: DependencyList,
) => AsyncTask<Result>;

export const useAsyncTaskTimeout: UseAsyncTaskTimeout;

// delay task

export type UseAsyncTaskDelay = (
  milliSeconds: number,
  deps: DependencyList,
) => AsyncTask<true>;

export const useAsyncTaskDelay: UseAsyncTaskDelay;

// fetch task

export type UseAsyncTaskFetch = <Result>(
  input: string | Request,
  init?: RequestInit,
  bodyReader?: (b: Body) => Promise<Result>,
) => AsyncTask<Result>;

export const useAsyncTaskFetch: UseAsyncTaskFetch;
export const useFetch: UseAsyncTaskFetch;

// axios task

export type UseAsyncTaskAxios = <
  Result,
  AxiosInstance = unknown,
  AxiosRequestConfig = unknown,
>(
  axios: AxiosInstance,
  config: AxiosRequestConfig,
  deps: DependencyList,
) => AsyncTask<Result>;

export const useAsyncTaskAxios: UseAsyncTaskAxios;
export const useAxios: UseAsyncTaskAxios;
export default useAsyncTaskAxios;

// wasm task

export type UseAsyncTaskWasm = <Result>(
  input: string | Request,
  importObject?: object,
) => AsyncTask<Result>;

export const useAsyncTaskWasm: UseAsyncTaskWasm;
export const useWasm: UseAsyncTaskWasm;
