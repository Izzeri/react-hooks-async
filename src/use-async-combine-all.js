import { useCallbackOne as useCallback } from 'use-memo-one';

import { useAsyncTask } from './use-async-task';

export const useAsyncCombineAll = (...asyncTasks) => {
  const task = useAsyncTask(useCallback(
    async (abortController) => {
      abortController.signal.addEventListener('abort', () => {
        asyncTasks.forEach((asyncTask) => {
          if (asyncTask.abort) {
            asyncTask.abort();
          }
        });
      });
      asyncTasks.forEach((asyncTask) => {
        if (!asyncTask.start) throw new Error('no asyncTask.start');
        asyncTask.start();
      });
    },
    // TODO Do we have a better way?
    // eslint-disable-next-line react-hooks/exhaustive-deps
    asyncTasks.map(({ start }) => start),
  ));
  return {
    ...task,
    pending: asyncTasks.some(({ pending }) => pending),
    error: asyncTasks.find(({ error }) => error),
    errorAll: asyncTasks.map(({ error }) => error),
    result: asyncTasks.map(({ result }) => result),
  };
};
