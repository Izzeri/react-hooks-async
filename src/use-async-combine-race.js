import { useCallback, useEffect, useMemo } from 'react';

import { useAsyncTask } from './use-async-task';
import { useMemoList } from './utils';

export const useAsyncCombineRace = (...asyncTasks) => {
  const memoAsyncTasks = useMemoList(asyncTasks, (a, b) => a.start === b.start);
  const task = useAsyncTask(useCallback(
    async (abortController) => {
      abortController.signal.addEventListener('abort', () => {
        memoAsyncTasks.forEach((asyncTask) => {
          asyncTask.abort();
        });
      });
      // start everything
      memoAsyncTasks.forEach((asyncTask) => {
        asyncTask.start();
      });
    },
    [memoAsyncTasks],
  ));
  const finishedIndex = asyncTasks.findIndex(({ pending }) => !pending);
  useEffect(() => {
    // if there's one task finished, abort all the others
    if (finishedIndex >= 0) {
      asyncTasks.forEach((asyncTask, i) => {
        if (i !== finishedIndex) {
          asyncTask.abort();
        }
      });
    }
  });
  const taskAborted = asyncTasks.every(({ aborted }) => aborted);
  const taskPending = asyncTasks.every(({ pending }) => pending);
  const taskError = asyncTasks.find(({ error }) => error);
  const taskErrorAll = useMemoList(asyncTasks.map(({ error }) => error));
  const taskResult = useMemoList(asyncTasks.map(({ result }) => result));
  return useMemo(() => ({
    start: task.start,
    abort: task.abort,
    started: task.started,
    aborted: taskAborted,
    pending: taskPending,
    error: taskError,
    errorAll: taskErrorAll,
    result: taskPending ? null : taskResult,
  }), [
    task.start,
    task.abort,
    task.started,
    taskAborted,
    taskPending,
    taskError,
    taskErrorAll,
    taskResult,
  ]);
};
