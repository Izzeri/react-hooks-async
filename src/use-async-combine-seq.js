import { useEffect } from 'react';
import {
  useCallbackOne as useCallback,
  useMemoOne as useMemo,
} from 'use-memo-one';

import { useAsyncTask } from './use-async-task';

// eslint-disable-next-line react-hooks/exhaustive-deps
const useMemoList = items => useMemo(() => items, items);

export const useAsyncCombineSeq = (...asyncTasks) => {
  const task = useAsyncTask(useCallback(
    async (abortController) => {
      abortController.signal.addEventListener('abort', () => {
        asyncTasks.forEach((asyncTask) => {
          asyncTask.abort();
        });
      });
      // start the first one
      asyncTasks[0].start();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    asyncTasks.map(({ start }) => start),
  ));
  useEffect(() => {
    // if prev task is finished, start next task
    const index = asyncTasks.findIndex(({ started }) => !started);
    const prevTask = asyncTasks[index - 1];
    const nextTask = asyncTasks[index];
    if (nextTask && prevTask && !prevTask.pending && !prevTask.error) {
      nextTask.start();
    }
  });
  const taskPending = asyncTasks.some(({ pending }) => pending);
  const taskError = asyncTasks.find(({ error }) => error);
  const taskErrorAll = useMemoList(asyncTasks.map(({ error }) => error));
  const taskResult = useMemoList(asyncTasks.map(({ result }) => result));
  return useMemo(() => ({
    start: task.start,
    abort: task.abort,
    started: task.started,
    pending: taskPending,
    error: taskError,
    errorAll: taskErrorAll,
    result: taskResult,
  }), [
    task.start,
    task.abort,
    task.started,
    taskPending,
    taskError,
    taskErrorAll,
    taskResult,
  ]);
};
