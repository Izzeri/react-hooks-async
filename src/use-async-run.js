import { useEffect } from 'react';

import { useMemoList } from './utils';

export const useAsyncRun = (asyncTask, ...args) => {
  const start = asyncTask && asyncTask.start;
  const started = asyncTask && asyncTask.started;
  const abort = asyncTask && asyncTask.abort;
  const memoArgs = useMemoList(args);
  useEffect(() => {
    if (start) {
      start(...memoArgs);
    }
  }, [start, memoArgs]);
  useEffect(() => {
    const cleanup = () => {
      if (started && abort) {
        abort();
      }
    };
    return cleanup;
  }, [started, abort]);
};
