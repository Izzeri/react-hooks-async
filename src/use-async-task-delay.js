import { useAsyncTask } from './use-async-task';

const createAbortError = (message) => {
  try {
    return new DOMException(message, 'AbortError');
  } catch (e) {
    const err = new Error(message);
    err.name = 'AbortError';
    return err;
  }
};

export const useAsyncTaskDelay = (milliSeconds, deps) => useAsyncTask(
  abortController => new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      resolve(true);
    }, milliSeconds);
    abortController.signal.addEventListener('abort', () => {
      clearTimeout(id);
      reject(createAbortError('timer aborted'));
    });
  }),
  deps,
);
