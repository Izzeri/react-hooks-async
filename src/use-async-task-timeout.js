import { useCallbackOne as useCallback } from 'use-memo-one';

import { useAsyncTask } from './use-async-task';

const createAbortError = (message) => {
  try {
    return new DOMException(message, 'AbortError');
  } catch (e) {
    const err = new Error(message);
    err.name = 'AbortError';
    return err;
  }
};

export const useAsyncTaskTimeout = (func, delay) => useAsyncTask(useCallback(
  abortController => new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      resolve(func());
    }, delay);
    abortController.signal.addEventListener('abort', () => {
      clearTimeout(id);
      reject(createAbortError('timer aborted'));
    });
  }),
  [func, delay],
));
