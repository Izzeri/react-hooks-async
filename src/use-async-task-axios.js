import { useCallbackOne as useCallback } from 'use-memo-one';

import { useAsyncTask } from './use-async-task';
import { useAsyncRun } from './use-async-run';

export const useAsyncTaskAxios = (axios, config) => useAsyncTask(useCallback(
  async (abortController, configOverride) => {
    const source = axios.CancelToken.source();
    abortController.signal.addEventListener('abort', () => {
      source.cancel('canceled');
    });
    try {
      return await axios({
        ...config,
        ...configOverride,
        cancelToken: source.token,
      });
    } catch (e) {
      if (axios.isCancel(e)) {
        const error = new Error(e.message);
        error.name = 'AbortError';
        throw error;
      }
      throw e;
    }
  },
  [axios, config],
));

export const useAxios = (...args) => {
  const asyncTask = useAsyncTaskAxios(...args);
  useAsyncRun(asyncTask);
  return asyncTask;
};
