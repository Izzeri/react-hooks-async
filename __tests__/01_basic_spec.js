import React, { StrictMode } from 'react';
import { render, cleanup } from '@testing-library/react';

import { useAsyncTask, useAsyncRun, useAsyncTaskTimeout } from '../src/index';

// FIXME we should fake timers
// jest.useFakeTimers();
const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

describe('basic spec', () => {
  afterEach(cleanup);

  it('should have a function', () => {
    expect(useAsyncTask).toBeDefined();
  });

  it('should create a component with timeout', async () => {
    const Waiting = ({ abort }) => (
      <div>
        Waiting...
        <button type="button" onClick={abort}>Abort</button>
      </div>
    );
    const renderHi = () => <h1>Hi</h1>;
    const DelayedMessage = ({ delay }) => {
      const asyncTask = useAsyncTaskTimeout(renderHi, delay, [renderHi, delay]);
      useAsyncRun(asyncTask);
      const {
        pending,
        error,
        result,
        abort,
      } = asyncTask;
      if (error) return <div>Error:{error.message}</div>;
      if (pending) return <Waiting abort={abort} />;
      return <div>Result:{result}</div>;
    };
    const App = () => (
      <StrictMode>
        <div>
          <DelayedMessage delay={300} />
          <DelayedMessage delay={100} />
        </div>
      </StrictMode>
    );
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    // jest.advanceTimersByTime(100);
    await sleep(100);
    expect(container).toMatchSnapshot();
    // jest.advanceTimersByTime(200);
    await sleep(200);
    expect(container).toMatchSnapshot();
  });
});
