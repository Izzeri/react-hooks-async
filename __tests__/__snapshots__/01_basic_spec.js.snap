// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create a component with timeout 1`] = `
<div>
  <div>
    <div>
      Waiting...
      <button
        type="button"
      >
        Abort
      </button>
    </div>
    <div>
      Waiting...
      <button
        type="button"
      >
        Abort
      </button>
    </div>
  </div>
</div>
`;

exports[`basic spec should create a component with timeout 2`] = `
<div>
  <div>
    <div>
      Waiting...
      <button
        type="button"
      >
        Abort
      </button>
    </div>
    <div>
      Result: 
      <h1>
        Hi
      </h1>
    </div>
  </div>
</div>
`;

exports[`basic spec should create a component with timeout 3`] = `
<div>
  <div>
    <div>
      Result: 
      <h1>
        Hi
      </h1>
    </div>
    <div>
      Result: 
      <h1>
        Hi
      </h1>
    </div>
  </div>
</div>
`;
